green_bit_llm/evaluation/evaluate.py

```diff
@@ -258,7 +274,97 @@ def main(args):
 
     eval_results = {"{}".format(args.model): eval_results}
 
-    add_dict_to_json_file(file_path="{}".format(os.path.join(args.save_dir, "eval_results.json")), new_data=eval_results)
+    add_dict_to_json_file(file_path="{}".format(os.path.join(args.save_dir, "eval_greenbit_engine_results.json")), new_data=eval_results)
+
+def evaluate_vllm(args):
+    logits_list = []
+    def forward_hook(module, input, output):
+        lm_head, hidden_states, sampling_metadata, *embedding_bias = input
+        embedding_bias = embedding_bias[0] if embedding_bias else None
+        logits = module._get_logits(hidden_states, lm_head, embedding_bias)
+        if logits is not None:
+            if module.soft_cap is not None:
+                logits = logits / module.soft_cap
+                logits = torch.tanh(logits)
+                logits = logits * module.soft_cap
+            if module.scale != 1.0:
+                logits *= module.scale
+            logits = _apply_logits_processors(logits, sampling_metadata)
+            logits_list.append(logits)
+        return output 
+    
+    @torch.no_grad()
+    def calculate_ppl(model, testenc, seqlen, device='cuda'):
+        nsamples = testenc.numel() // seqlen
+        nlls = []
+        
+        sampling_params = SamplingParams(
+            temperature=1.0,
+            max_tokens=1,
+            logprobs=None
+        )
+        
+        for i in tqdm(range(nsamples)):
+            logits_list.clear()
+            batch = testenc[:, (i * seqlen):((i + 1) * seqlen)]
+            outputs = model.generate(prompts=None, prompt_token_ids=batch.tolist(), sampling_params=sampling_params)
+            logits = logits_list[0].to(device)
+            logits = logits.unsqueeze(0)
+            shift_logits = logits[:, :-1, :]
+            shift_labels = testenc[:, (i * seqlen): ((i + 1) * seqlen)][
+                            :, 1:
+                            ].to(device)
```

Comment:
Please adjust the styling (see https://peps.python.org/pep-0008/#indentation or follow [black style](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html)).